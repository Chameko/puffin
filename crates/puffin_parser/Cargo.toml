[package]
name = "puffin_parser"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
puffin_ast = { path = "../puffin_ast" }
puffin_vfs = { path = "../puffin_vfs"}
puffin_error = { path = "../puffin_error" }
puffin_hir = { path = "../puffin_hir" }
salsa = "0.16"
rowan = "0.15"

[dev-dependencies]
insta = { version = "1.29" }

[profile.dev.package.insta]
opt-level = 3

[profile.dev.package.similar]
opt-level = 3
