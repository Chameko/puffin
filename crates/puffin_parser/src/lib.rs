/// Lexer for puffin
pub mod lexer;
/// Parser for parsing the Concrete Syntax Tree to an Abstract Syntax Tree
pub mod ast_parser;
/// Parser for parsing tokens into the Concrete Syntax Tree
pub mod parser;
