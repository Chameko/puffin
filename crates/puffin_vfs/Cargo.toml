[package]
name = "puffin_vfs"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
crossbeam-channel = "0.5.8"
fxhash = "0.2"
notify = "5.1.0"
relative-path = { version = "1.8", default-features = false }
thiserror = "1.0"
