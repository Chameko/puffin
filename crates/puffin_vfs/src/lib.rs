/// Contains the path interner
mod path_interner;
use path_interner::PathInterner;

/// Contains the moniter
mod monitor;
pub use monitor::Monitor;
pub use monitor::ForegroundMessage;
pub use monitor::MonitorMessage;

/// Contains the path types used in Puffin
pub mod path;
pub use path::AbsPath;
pub use path::AbsPathBuf;

use relative_path::RelativePath;
use thiserror::Error;

/// Identifier for a file in the virtual file system
#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Eq, Ord)]
#[repr(transparent)]
pub struct FileID(pub u32);

/// Error type for the virtual file system
#[derive(Error, Debug)]
pub enum VFSError {
    #[error("expected an absolute path")]
    NotAbsolutePath,
}

/// The virtual file system
#[derive(Debug)]
pub struct VFS {
    /// The internal path interner
    interner: PathInterner,
}

impl VFS {
    /// Either adds a path and returns the [`FileID`] or returns the [`FileID`] if it already exists
    pub fn intern(&mut self, path: &AbsPath) -> FileID {
        self.interner.intern(path)
    }

    /// Get an absolute path from a file ID
    pub fn get_path(&self, id: FileID) -> Option<&AbsPath> {
        self.interner.get_path(id)
    }

    /// Intern a [RelativePath] that is relative to a supplied [AbsPath]
    pub fn intern_relative_path(&mut self, path: &RelativePath, relative_to: &AbsPath) -> FileID {
        self.interner.intern(&AbsPathBuf::try_from(path.to_logical_path(relative_to)).unwrap())
    }
}
