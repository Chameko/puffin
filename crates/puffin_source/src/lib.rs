pub mod module;
pub mod id;

use puffin_vfs::VFS;
pub use puffin_vfs::FileID;
use relative_path::RelativePathBuf;
use std::sync::Arc;
use crate::module::{ModuleTree, generate_module_tree};

/// Type use for marking slices of text in the source files
pub type TextSlice = std::ops::RangeInclusive<u32>;

/// The query group for interacting with the source files
#[salsa::query_group(SourceStorage)]
pub trait SourceDatabase {
    /// Set the source tree
    #[salsa::input]
    fn source_tree(&self) -> Arc<SourceTree>;

    /// Set the [VFS]
    #[salsa::input]
    fn vfs(&self) -> Arc<VFS>;

    #[salsa::invoke(generate_module_tree)]
    fn generate_module_tree(&self) -> Arc<ModuleTree>;

    /// Get the relative path of a file to the source tree
    fn relative_path(&self, file_id: FileID) -> Option<RelativePathBuf>;
}

/// Get the relative path of a file to the source tree
fn relative_path(db: &dyn SourceDatabase, file_id: FileID) -> Option<RelativePathBuf> {
    let tree = db.source_tree();
    let vfs = db.vfs();
    tree.get_relative_path(file_id, vfs)
}

/// The source tree represents a tree of sources in a Puffin project. Each file is relative to the source tree
#[derive(Debug, Clone)]
pub struct SourceTree {
    /// The base directory of the source tree. Everything else is relative to this
    dir: FileID,
    /// The sources in the source tree
    pub sources: Vec<Source>
}

impl SourceTree {
    /// Find a source in the source tree
    pub fn find_source(&self, file_id: FileID) -> Option<&Source> {
        for source in &self.sources {
            if source.file == file_id {
                return Some(source)
            }
        }
        None
    }

    /// Get the path of a source relative to this source tree
    pub fn get_relative_path(&self, file_id: FileID, vfs: Arc<VFS>) -> Option<RelativePathBuf> {
        let src_path = self.find_source(file_id).map(|s| vfs.get_path(s.file)).flatten();
        let path = vfs.get_path(self.dir);
        if let (Some(src_path), Some(path)) = (src_path, path) {
            if let Ok(path) = src_path.strip_prefix(path) {
                Some(RelativePathBuf::from(path.as_os_str().to_string_lossy().to_string()))
            } else {
                None
            }
        } else {
            None
        }
    }
}


/// A source in the source tree
#[derive(Debug, Clone, PartialEq)]
pub struct Source {
    // The file that the source belongs to
    pub file: FileID,
    // The text contained within the file
    pub text: String,
}

impl Source {
    /// Get the text of a source file
    pub fn get_text(&self) -> &str {
        self.text.as_str()
    }
}
