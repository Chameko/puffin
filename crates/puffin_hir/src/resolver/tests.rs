use std::{path::PathBuf, sync::Arc};

use puffin_parser::{lexer::LexerStorage, parser::{ParserStorage, ParserDatabase}};
use puffin_source::{Source, SourceTree, SourceStorage, SourceDatabase};
use crate::{model::{InternStorage, InternDatabase, Function, Stmt, Expr}, def::{DefStorage, DefDatabase}, resolver::{ResolveStorage, ResolveDatabase}, item_tree::SplitItemTreeNode, id::{StmtID, ExprID}};
use puffin_vfs::AbsPathBuf;

use super::Resolved;

#[salsa::database(SourceStorage, LexerStorage, ParserStorage, InternStorage, DefStorage, ResolveStorage)]
#[derive(Default)]
struct HirTest {
    storage: salsa::Storage<Self>
}

impl salsa::Database for HirTest {}

/// Parses the inputed string and creates a snapshot
fn standard_test(src: &str) -> String {
    let mut vfs = puffin_vfs::VFS::new();
    let file_id = vfs.intern(&AbsPathBuf::try_from(PathBuf::from("/test.pf")).unwrap());
    let dir = vfs.intern(&AbsPathBuf::try_from(PathBuf::from("/")).unwrap());
    let source = Source{ file: file_id, text: src.to_string() };
    let src_tree = SourceTree::new(dir, vec![source]);
    let mut hir_test = HirTest::default();
    hir_test.set_source_tree(Arc::new(src_tree));
    hir_test.set_vfs(Arc::new(vfs));
    let item_tree = hir_test.item_tree(file_id);
    let funcs = item_tree.functions();
    let mut main = None;
    for func in funcs {
        if item_tree[func].signature.name == "main" {
            main = Some(hir_test.intern_function(Function::to_sig_id(func)));
        }
    }
    if let Some(main) = main {
        let mut resolved = hir_test.resolve_query(main);
        let mut output = String::new();
        for error in resolved.diagnostics {
            output.push_str(&format!("{}\n", error.display(&hir_test.source_tree(), &hir_test.vfs())));
        }
        output.push_str(&output_types(resolved.resolved.pop().unwrap(), src, hir_test));
        output
    } else {
        panic!("No main function")
    }
}

/// Outputs the types of various statements and expressions
fn output_types(res: Resolved, src: &str, hir_test: HirTest) -> String {
    let mut output = output_stmt(res.resolved_body.source, &res, src, 0);
    for error in res.diagnostics {
        std::env::set_var("NO_COLOR", "true");
        let error_string = error.display(&hir_test.source_tree(), &hir_test.vfs());
        println!("{}", error_string);
        output.push_str(&error_string);
    }
    output
}

/// Used to display various statements and their infered types
fn output_stmt(stmt: StmtID, res: &Resolved, src: &str, mut indent: usize) -> String {
    let slice = res.resolved_body_src.stmt_map[stmt].element.text_slice();
    let mut output = format!("{}[[ {} ]]\n", " ".repeat(2 * indent), &src[*slice.start() as usize..=*slice.end() as usize].trim());
    indent += 1;
    match &res.resolved_body.stmt_alloc[stmt] {
        Stmt::ExprStmt(e) => {
            output_expr(*e, res, src, indent)
        },
        Stmt::Let { ty, expr } => {
            let pat_name = match &res.resolved_body.pat_alloc[ty.pat] {
                crate::model::Pattern::Ident(i) => i.name.to_string(),
                _ => String::new(),
            };
            output.push_str(&" ".repeat(2 * indent));
            output.push_str(&format!("{} => {:?}", pat_name, res.resolved_body.type_alloc[ty.ty].display(&res.resolved_body.type_alloc)));
            if let Some(expr) = expr {
                output.push_str(&format!(" from {{\n{}{}}}\n", output_expr(*expr, res, src, indent), " ".repeat(2*indent)))
            }
            output
        },
        Stmt::While { condition, exec } => {
            output.push_str(&" ".repeat(2 * indent));
            output.push_str(&output_expr(*condition, res, src, indent));
            output.push_str(&" ".repeat(2 * indent));
            output.push_str(&output_stmt(*exec, res, src, indent));
            output
        },
        Stmt::If { condition, truthy, falsey } => {
            output.push_str(&" ".repeat(2 * indent));
            output.push_str(&output_expr(*condition, res, src, indent));
            output.push_str(&" ".repeat(2 * indent));
            output.push_str(&output_stmt(*truthy, res, src, indent));
            if let Some(falsey) = falsey {
            output.push_str(&" ".repeat(2 * indent));
                output.push_str(&output_stmt(*falsey, res, src, indent));
            }
            output
        },
        Stmt::Block { stmts } => {
            for stmt in stmts {
                output.push_str(&" ".repeat(2 * indent));
                output.push_str(&output_stmt(*stmt, res, src, indent));
            }
            output
        },
        _ => String::new(),
    }
}

/// Used to display various expressions and their infered types
fn output_expr(expr: ExprID, res: &Resolved, src: &str, mut indent: usize) -> String {
    let slice = res.resolved_body_src.expr_map[expr].element.text_slice();
    indent += 1;
    let mut output = format!("{}{{ {} }} => ", " ".repeat(2 * indent), &src[*slice.start() as usize..=*slice.end() as usize].trim());
    match &res.resolved_body.expr_alloc[expr] {
        Expr::Binary { lhs, rhs, ty, ..} => {
            output.push_str(&format!("{:?}\n", res.resolved_body.type_alloc[*ty].display(&res.resolved_body.type_alloc)));
            output.push_str(&output_expr(*lhs, res, src, indent));
            output.push_str(&output_expr(*rhs, res, src, indent));
            output
        },
        Expr::Prefix { expr, ty, ..} => {
            output.push_str(&format!("{:?}\n", res.resolved_body.type_alloc[*ty].display(&res.resolved_body.type_alloc)));
            output.push_str(&output_expr(*expr, res, src, indent));
            output
        },
        Expr::Assign { assignee, assign_to, ty } => {
            output.push_str(&format!("{:?}\n", res.resolved_body.type_alloc[*ty].display(&res.resolved_body.type_alloc)));
            output.push_str(&output_expr(*assignee, res, src, indent));
            output.push_str(&output_expr(*assign_to, res, src, indent));
            output
        },
        Expr::Pattern(p) => {
            output.push_str(&format!("{:?}\n", res.resolved_body.pat_alloc[*p]));
            output
        }
        Expr::Paren(p) => {
            output.push_str(&format!("({:?})\n", output_expr(*p, res, src, indent)));
            output
        }
        Expr::Missing => {
            output.push_str("MISSING\n");
            output
        }
    }
}

#[test]
fn basic_func() {
    let src = "fun main() {\nlet a = 1\nlet b = 2\nlet c = a + b\n}";
    insta::assert_snapshot!(standard_test(src));
}

#[test]
fn func_with_param() {
    let src = "fun main(a: int, b: int) {\nlet c = a + b\n}";
    insta::assert_snapshot!(standard_test(src));
}

#[test]
fn incorrect_types() {
    let src = "fun main() {\nlet a = 1\nlet b = 2.0\nlet c = a + b\n}";
    insta::assert_snapshot!(standard_test(src));
}

#[test]
fn incorrect_param() {
    let src = "fun main(a: float, b: int) {\nlet c = a + b\n}";
    insta::assert_snapshot!(standard_test(src));
}
