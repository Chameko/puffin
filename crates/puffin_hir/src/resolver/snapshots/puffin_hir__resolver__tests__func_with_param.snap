---
source: crates/puffin_hir/src/resolver/tests.rs
expression: standard_test(src)
---
[[ {
let c = a + b
} ]]
    [[ let c = a + b ]]
    c => "Concrete(Int)" from {
      { a + b } => "FunctionType { Concrete(Int), Concrete(Int), ret: Concrete(Int) }"
        { a } => Ident(Ident { name: "a" })
        { b } => Ident(Ident { name: "b" })
    }

